runtime/src/account_rent_state.rs

```diff
@@ -24,35 +26,54 @@ impl RentState {
     pub(crate) fn from_account(account: &AccountSharedData, rent: &Rent) -> Self {
         if account.lamports() == 0 {
             Self::Uninitialized
-        } else if !rent.is_exempt(account.lamports(), account.data().len()) {
-            Self::RentPaying(account.data().len())
-        } else {
+        } else if rent.is_exempt(account.lamports(), account.data().len()) {
             Self::RentExempt
+        } else {
+            Self::RentPaying {
+                data_size: account.data().len(),
+                lamports: account.lamports(),
+            }
         }
     }
 
-    pub(crate) fn transition_allowed_from(&self, pre_rent_state: &RentState) -> bool {
-        if let Self::RentPaying(post_data_size) = self {
-            if let Self::RentPaying(pre_data_size) = pre_rent_state {
-                post_data_size == pre_data_size // Cannot be RentPaying if resized
-            } else {
-                false // Only RentPaying can continue to be RentPaying
+    pub(crate) fn transition_allowed_from(
+        &self,
+        pre_rent_state: &RentState,
+        restrict_rent_paying_accounts_ttl_extension: bool,
+    ) -> bool {
+        match self {
+            Self::Uninitialized | Self::RentExempt => true,
+            Self::RentPaying {
+                data_size: post_data_size,
+                lamports: post_lamports,
+            } => {
+                match pre_rent_state {
+                    Self::Uninitialized | Self::RentExempt => false,
+                    Self::RentPaying {
+                        data_size: pre_data_size,
+                        lamports: pre_lamports,
+                    } => {
+                        // Cannot be RentPaying if resized
+                        post_data_size == pre_data_size && {
+                            !restrict_rent_paying_accounts_ttl_extension
+                                || post_lamports <= pre_lamports
+                        }
```

Comment:
Did `cargo fmt` wrap this extended boolean in brackets like this instead of parens? That's weird.

Reply:
Rust accepts both; I think in this case `{` gives more readable format:
```rust
post_data_size == pre_data_size && {
    !prevent_crediting_rent_paying_accounts || post_lamports <= pre_lamports
}
```
vs
```rust
post_data_size == pre_data_size
    && (!prevent_crediting_rent_paying_accounts
        || post_lamports <= pre_lamports)
```